### ISSUE TITLE
bashbrew: Generic support for Docker Registries

### ISSUE THREAD

**Issue by jsok** [docker-library/official-images#7854]
Right now the tool depends on the [DockerHub API](https://github.com/docker-library/official-images/blob/936ab74bae3cc42e3dc8d6bec01a454fd4b81af6/bashbrew/go/src/bashbrew/hub.go#L34) which isn't compatible with the [Docker Registry HTTP API V2](https://docs.docker.com/registry/spec/api/).

This makes `bashbrew` less portable and not suited for managing docker libraries running on other registries.

Would you be open to supporting patches that:
1. Provide config flag(s) to instruct `bashbrew` to use the official registry API for tag metadata scraping
1. Add the necessary authentication support to use the registry API

**Comment by tianon** [comment]
This was discussed a little bit previously in https://github.com/docker-library/official-images/pull/4637#issuecomment-408201141, but that discussion had a lot of extra content so I'll copy the salient points here:

> the Docker Hub API code is merely an optimization -- it should error out and be ignored fairly quickly if pushing to a non-Hub registry via `--namespace`

> So, the reason we implemented this in the first place was because Docker Hub is slow (not because we really needed to know when the push would be a no-op).  I still think this would be better implemented for non-slow registries outside `bashbrew` (especially since the alternative is to pull in `docker/distribution` code to properly parse the registry value out of a given image name, and that's a bit heavier than I'd like to add here -- even implementing the registry protocol is a bit heavier than I'd like `bashbrew` to be without strong reason, not to mention maintaining functionality we then won't be using or testing).

**Comment by jsok** [comment]
@tianon thanks for the link, I promise I did search for similar issues but it's not easy due to the commit noise of the everyday library updates 😄 

Happy with that explanation, my private registry should be sufficiently quick for no-op pushes so happy to close this.

Before I go and create another issue, what would your stance be on adding Git auth support? My org's Github Enterprise installation has anonymous clones disabled site-wide. I've got a patch to add netrc support, the integration is pretty straightforward (excluding the netrc parsing impl):

```go
	netrcClient := &http.Client{Transport: &netrcTransport{}}
	goGitClient.InstallProtocol("https", goGitHttp.NewClient(netrcClient))
```

It's a no-op if a netrc doesn't exist or if the domain isn't found in the file.
It would require adding a dep to do the netrc parsing though.
